frontend/src/views/FuelCodes/AddFuelCode/AddEditFuelCode.jsx

```diff
@@ -0,0 +1,373 @@
+  const {
+    data: existingFuelCode,
+    isLoading: isLoadingExistingCode,
+    refetch
+  } = useGetFuelCode(fuelCodeID)
+
+  useEffect(() => {
+    if (optionsData) {
+      const updatedColumnDefs = fuelCodeColDefs(
+        optionsData,
+        errors,
+        !existingFuelCode,
+        existingFuelCode?.fuelCodeStatus.status !==
+          FUEL_CODE_STATUSES.APPROVED && hasRoles(roles.analyst)
+      )
+      setColumnDefs(updatedColumnDefs)
+    }
+  }, [errors, optionsData, existingFuelCode])
+
+  useEffect(() => {
+    if (existingFuelCode) {
+      setRowData([existingFuelCode])
+    } else {
+      setRowData([
+        {
+          id: uuid(),
+          prefixId: 1,
+          fuelSuffix: optionsData?.fuelCodePrefixes?.find(
+            (item) => item.prefix === 'BCLCF'
+          ).nextFuelCode
+        }
+      ])
+    }
+  }, [optionsData, existingFuelCode, isGridReady])
+
+  const onGridReady = (params) => {
+    setGridReady(true)
+    params.api.sizeColumnsToFit()
+  }
+
+  const onCellValueChanged = useCallback(
+    async (params) => {
+      const updatedData = { ...params.data, modified: true }
+
+      if (params.colDef.field === 'prefix') {
+        updatedData.fuelSuffix = optionsData?.fuelCodePrefixes?.find(
+          (item) => item.prefix === params.newValue
+        ).nextFuelCode
+      }
+
+      params.api.applyTransaction({ update: [updatedData] })
+    },
+    [optionsData?.fuelCodePrefixes]
+  )
+
+  const handleDeleteFuelCode = async (fuelCodeId, params) => {
+    if (fuelCodeId) {
+      try {
+        await deleteFuelCode(fuelCodeId)
+        params.api.applyTransaction({ remove: [params.node.data] })
+        alertRef.current?.triggerAlert({
+          message: 'Row deleted successfully.',
+          severity: 'success'
+        })
+      } catch (error) {
+        console.error('Error deleting row:', error)
+        alertRef.current?.triggerAlert({
+          message: `Error deleting row: ${error.message}`,
+          severity: 'error'
+        })
+      }
+    } else {
+      params.api.applyTransaction({ remove: [params.node.data] })
+    }
+    setModalData(null)
+  }
+
+  const handleApproveDraftCode = async (fuelCodeId) => {
+    await approveFuelCode(fuelCodeId)
+    await refetch()
+    setModalData(null)
+  }
+
+  const openDeleteModal = (fuelCodeId, params) => {
+    setModalData({
+      primaryButtonAction: () => handleDeleteFuelCode(fuelCodeId, params),
+      primaryButtonText: t('fuelCode:deleteFuelCodeBtn'),
+      secondaryButtonText: t('cancelBtn'),
+      title: t('fuelCode:deleteFuelCode'),
+      content: (
+        <Stack>
+          <BCTypography variant="h6">
+            {t('fuelCode:deleteFuelCode')}
+          </BCTypography>
+          <BCTypography mt={1} variant="body5">
+            {t('fuelCode:deleteConfirmText')}
+          </BCTypography>
+        </Stack>
+      )
+    })
+  }
+
+  const openApprovalModal = async (fuelCodeId) => {
```

Comment:
One optimization to help limit re-renders 
```jsx
const handleOpenApprovalModal = useCallback(() => {
  openApprovalModal(fuelCodeID)
}, [fuelCodeID])

<BCButton
  onClick={handleOpenApprovalModal}
>

Reply:
Thanks! Added!